main/src/main/scala/sbt/internal/server/BuildServerReporter.scala

```diff
@@ -117,13 +127,18 @@ final class BuildServerReporterImpl(
       }
     }
   }
-
   override def sendFailureReport(sources: Array[VirtualFile]): Unit = {
     for {
       source <- sources
       filePath <- toSafePath(source)
     } {
-      val diagnostics = problemsByFile.getOrElse(filePath, Vector())
+      val diagnostics = problemsByFile.getOrElse(filePath, Vector.empty)
+
+      // mark file as file with problems
+      if (diagnostics.nonEmpty) {
+        bspCompileState.hasAnyProblems.add(filePath)
+      }
```

Comment:
What if the diagnostics are empty and `hasAnyProblems` is false for this `filePath`. Maybe we can not send any report this time.

```suggestion
    val hadProblems = bspCompileState.hasAnyProblems.remove(filePath)
    
    val shouldPublish = hadProblems || diagnostics.nonEmpty || shouldReportAllProblems
    
    if (shouldPublish) {
      if (diagnostics.nonEmpty) {
        bspCompileState.hasAnyProblems.add(filePath)
      }
      
      ...
```

Reply:
Wanted to write a testcase for this, but I couldn't because the following flow is broken:
- project is compiling
- introduce error
- reload workspace (this seems to create a new bsp reporter which will have `compiledAtLeastOnce` set to false)
- compile project - because `compiledAtLeastOnce` is set to false all diagnostics will be returned